contracts/truefi2/CrvBaseRateOracle.sol

```diff
@@ -0,0 +1,157 @@
+     * avgRate = -------------------------------------------------------------------------------------------
+     *                                                  (t - t_0)
+     *
+     * where v_i, t_i are values of the prices and their respective timestamps
+     * stored in the historical buffer, v is a value of current price ant t is its timestamp.
+     * Index n-1 corresponds to the most recent values and index 0 to the oldest ones.
+     * Notice that whether we are going to use the whole buffer or not
+     * depends on value of timeToCover parameter.
+     * @param timeToCover For how much time average should be calculated
+     * @return Average rate in basis points
+     */
+    function calculateAverageRate(uint256 timeToCover) public view returns (uint256) {
+        require(1 days <= timeToCover && timeToCover <= 365 days, "CrvBaseRateOracle: Expected amount of time in range 1 to 365 days");
+        // estimate how much buffer we need to use
+        uint256 bufferSizeNeeded = timeToCover.div(cooldownTime);
+        require(bufferSizeNeeded <= bufferSize(), "CrvBaseRateOracle: Needed buffer size cannot exceed size limit");
+        uint16 iidx = histBuffer.insertIndex;
+        uint256 sum;
+        uint256 totalTime;
+        for (uint16 i = 1; i < bufferSizeNeeded; i++) {
+            uint16 prevIdx = uint16(iidx.add(bufferSize()).sub(i).sub(1) % bufferSize());
+            if (histBuffer.timestamps[prevIdx] == 0) {
+                break;
+            }
+            uint16 idx = uint16(prevIdx.add(1) % bufferSize());
+            uint256 dt = histBuffer.timestamps[idx].sub(histBuffer.timestamps[prevIdx]);
+            sum = sum.add(histBuffer.baseRates[idx].add(histBuffer.baseRates[prevIdx]).mul(dt));
+            totalTime = totalTime.add(dt);
+        }
+        uint256 curCrvBaseRate = curve.get_virtual_price();
+        uint256 curTimestamp = block.timestamp;
+        uint16 idx = uint16(iidx.add(bufferSize()).sub(1) % bufferSize());
+        sum = sum.add(curCrvBaseRate.add(histBuffer.baseRates[idx]).mul(curTimestamp.sub(histBuffer.timestamps[idx])));
+        totalTime = totalTime.add(curTimestamp.sub(histBuffer.timestamps[idx]));
+        return sum.mul(10000).div(2).div(totalTime);
+    }
+
+    /**
+     * @dev Profit in 7 days based on average rate
+     * from collected data from last 7 days
+     * @return Calculated weekly profit
+     */
+    function weeklyProfit() public view returns (int256) {
+        int256 avgRate = int256(calculateAverageRate(7 days));
+        uint256 curCrvBaseRate = curve.get_virtual_price();
+        return ((int256(curCrvBaseRate.mul(10000)) - avgRate) * 10000) / avgRate;
+    }
+
+    /**
+     * @dev Profit in 30 days based on average rate
+     * from collected data from last 30 days
+     * @return Calculated monthly profit
+     */
+    function monthlyProfit() public view returns (int256) {
```

Comment:
These three functions are almost identical; factor out the identical parts into an internal helper function.